core/store/migrate/migrations/0259_add_solana_schema.sql

```diff
@@ -14,11 +14,22 @@ CREATE TABLE solana.log_poller_filters (
     event_sig BYTEA NOT NULL,
     starting_block BIGINT NOT NULL,
     event_idl TEXT,
-    sub_key_paths TEXT[][], -- A list of subkeys to be indexed, represented by their json paths in the event struct
+    sub_key_paths json, -- A list of subkeys to be indexed, represented by their json paths in the event struct. Forced to use json's 2d array as text[][] requires all paths to have equal length.
     retention BIGINT NOT NULL DEFAULT 0, -- we don’t have to implement this initially, but good to include it in the schema
     max_logs_kept BIGINT NOT NULL DEFAULT 0 -- same as retention, no need to implement yet
 );
 
+-- This generates a unique BIGINT for the log_poller_filters table from hashing (name, chain_id, address, event_sig, sub_key_paths).
+-- Using an ordinary multi-column index on 5 columns would require a lot of extra storage space.
+-- Note for updating this if and when we drop support for postgresql 12 & 13: hashrecordextended() can be used directly in postgresql 14, avoiding the need for a helper function.
+-- The helper function is necessary only for the IMMUTABLE keyword.
+CREATE OR REPLACE FUNCTION solana.f_log_poller_filter_hash(name TEXT, chain_id TEXT, address BYTEA, event_sig BYTEA, sub_key_paths json)
+    RETURNS BIGINT
+    LANGUAGE SQL IMMUTABLE COST 25 PARALLEL SAFE AS 'SELECT hashtextextended(textin(record_out(($1,$2,$3,$4,$5))), 0)';
+
+CREATE UNIQUE INDEX IF NOT EXISTS solana_log_poller_filters_hash_key
+    ON solana.log_poller_filters (solana.f_log_poller_filter_hash(name, chain_id, address, event_sig, sub_key_paths));
```

Comment:
Good idea!  I probably wasn't very clear on what the name column should be in the design doc, but I was thinking we'd generate a unique name for each row--so we shouldn't include it in this hash function.

I think we want two UNIQUE constraints:

- Each row must have a unique value in the `name` column
- Each row must have a unique combination of `(chain_id, event_sig, address, sub_key_paths)`

In the evm LogPoller, there should also be only one name per filter, but because the same filters is allowed to have multiple addresses and multiple event sigs one filter ends up getting spread across multiple rows. This made a lot of queries more messy and complex. Since the multi-address & multi event filter feature wasn't used much anyway, I decided it's more trouble than it's worth on Solana and other chains going forward.  The plugin can always regsiter multiple filters if it needs to track more than one event per address which is what the plugins usually do anyway.

I'm a little unsure about what to do with the `event_name` column. That should have a 1-to-1 correspondence with `event_sig` it's just there for readability in case we need it. We could still consider eliminating one or the other of those columns, for now I guess let's leave them both and not worry about constraints on `event_name`.

Reply:
Oh, and I can say a bit more about this in case you're wondering what I had in mind:
> I was thinking we'd generate a unique name for each row

For evm it's left up to the product plugins to generate unique names, but it's not strictly enforced which makes the whole thing a bit risky.  A name collision between 2 filters, whether or not they are from the same plugin, could cause problems. To improve on that, Solana's RegisterFilter will append a random string onto the end of whatever name is passed in by the plugin. So that will then server as a unique identifier for that filter, but it's still something we can print out in logs that will be readable instead of just an id.

Reply:
Oh, actually I just realized I think we don't want this constraint (the second one) after all.

The name uniqueness constraint should be there, so that if someone tries to re-register the same filter we return an error. (We could also update the row on conflict, which I think is what we do for evm... but I think it's safer to just return an error if that happens, unless all of the fields are the same... so we avoid accidental collisions.)

But it should be allowed for two different filters to have the rest of the fields all identical if they want.  The id and name are the only things which need to be unique, and id is enforced automatically.

I was thinking you added this index based on the Solana design doc.  But now I'm realizing you probably just copied it from evm (no wonder it looks familiar!)--we shouldn't need it in this case.

Reply:
Yes, I've copied it from the evm implementation as design doc did not specify any uniqueness indexes for the filter except id (or I've missed that part).

I was under the impression that products are allowed to reregister filters and they may use it to update retention and other filter configurations.